File: Data Structures & Algorithms/count-paths/submission-3.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        row = [0]*(n + 1)
        row[n - 1] = 1
        for i in range(m - 1, -1, -1): 
            for j in range(n - 1, -1, -1):
                row[j] = row[j] + row[j + 1]

        return row[0]

File: Data Structures & Algorithms/search-2d-matrix/submission-3.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int t = 0, d = matrix.size() - 1;
        while (t <= d){
            int m = (t + d) / 2;
            if (matrix[m][0] > target){
                d = m - 1;
            } else if (matrix[m].back() < target){
                t = m + 1;
            } else {
                // start search within column
                int l = 0, r = matrix[0].size() - 1;
                while (l <= r){
                    int middle = (r + l) / 2;
                    if (matrix[m][middle] < target){
                        l = middle + 1;
                    } else if (matrix[m][middle] > target){
                        r = middle - 1;
                    } else{
                        return true;
                    }
                }
                return false; // target not found in this row, but it should be in this row
            }
        }
        return false;
    }
};
